packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -1054,9 +1104,39 @@ class EditableText extends StatefulWidget {
   /// {@template flutter.widgets.editableText.inputFormatters}
   /// Optional input validation and formatting overrides.
   ///
-  /// Formatters are run in the provided order when the text input changes. When
-  /// this parameter changes, the new formatters will not be applied until the
-  /// next time the user inserts or deletes text.
+  /// Formatters are run in the provided order when the user changes the text
+  /// contained in the widget. They're not applied when the changes are
+  /// selection only, or not initiated by the user.
+  ///
+  /// When this widget rebuilds, each input formatter in the new widget's
+  /// [inputFormatters] list checks the configuration of the input formatter
+  /// from the same location in the old [inputFormatters], to determine if the
+  /// new formatters need to be re-applied to the current [TextEditingValue] of
+  /// this widget.
+  ///
+  /// {@tool snippet}
+  ///
+  /// The following code uses a combination of 2 [TextInputFormatter]s and a
+  /// fictional `UsPhoneNumberFormatter` (which simply adds parentheses and
+  /// hypens), to turn user input into a valid United States telephone number,
+  /// in the format of (123)456-7890.
+  ///
+  /// The combined effect of the 3 formatters is idempotent, meaning applying it
+  /// to an already formatted value is a no-op. This is a required property for
+  /// user-supplied formatters. The fictional `UsPhoneNumberFormatter` is not
+  /// idempotent, thus should not be used by itself.
```

Comment:
Is UsPhoneNumberFormatter defined somewhere?  I see it at the top of the file, but it doesn't actually have any logic for adding parenthesis and hyphens.

I wanted to see it because I thought "idempotent" might be misused here.  If "4155555555" always produces "(415)555-5555" then it is idempotent I think.  Even if "(415)555-5555" produces something different: "((41)5)5-55-5555".  Maybe I'm misunderstanding or making assumptions, though.

This is related to the comment I left on line 38 of text_formatter.dart above.

Reply:
No it's fictional, I took the example from the gallery app. I found this formatter surprisingly difficult to implement (mostly because selection/composing region handling) and I wanted to keep the example short so I left it out. The formatter in the gallery app simply adds parentheses and hypens at fixed locations whenever it runs (so yeah it produces "((41)5)5-55-5555" if you run it twice in a row). I'll put a simplified `UsPhoneNumberFormatter` here. But again I found formatters hard to implement in general since it has to deal with selections and compositions.